### PR TITLE
Implement ClassPrivateMember Structure

### PR DESCRIPTION
**Pull request by ksh8281**

### CODE REVIEW

File: src/interpreter/ByteCodeInterpreter.cpp [hunk 1]
```diff
@@ -2772,21 +2772,37 @@ NEVER_INLINE void ByteCodeInterpreter::initializeClassOperation(ExecutionState&
...
+        Optional<Object*> outerClassConstructor;
+        auto home = state.mostNearestHomeObject();
+        if (home) {
+            if (home->isScriptClassConstructorPrototypeObject()) {
+                home = home->asScriptClassConstructorPrototypeObject()->constructor();
+            }
+            outerClassConstructor = home;
+        }
+
+
```
Comment (bbrto21) [inline src/interpreter/ByteCodeInterpreter.cpp:2784]: This algorithm is very similar to the implementation of ```ExecutionState::findPrivateMemberContextObject```. Can you make it more efficient?

File: src/runtime/Object.cpp [hunk 2]
```diff
@@ -1655,100 +1657,149 @@ bool Object::setIndexedProperty(ExecutionState& state, const Value& property, co
...
-void Object::privateFieldAdd(ExecutionState& state, AtomicString propertyName, const Value& value)
+static ObjectPrivateMemberDataChain* ensurePieceOnPrivateMemberChain(ExecutionState& state, ObjectExtendedExtraData* e, Object* contextObject)
 {
-    auto& v = ensureObjectExtendedExtraData()->m_privateFieldValues;
-    for (size_t i = 0; i < v.size(); i++) {
-        if (UNLIKELY(v[i].first == propertyName)) {
-            ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "Cannot add private field %s with same name twice", propertyName.string());
+    Optional<ObjectPrivateMemberDataChain*> piece = e->m_privateMemberChain;
+    if (e->m_privateMemberChain) {
+        while (true) {
```
Comment (bbrto21) [inline src/runtime/Object.cpp:1664]: How about re-write a while statement using the expression below?

```suggestion
        while (piece) {
```

Comment (ksh8281) [reply src/runtime/Object.cpp:1664]: piece is always true.

File: src/interpreter/ByteCodeInterpreter.cpp (hunk 1 above)
Comment (ksh8281) [reply src/interpreter/ByteCodeInterpreter.cpp:2784]: Nice catch. I added some util function for that

File: src/runtime/Object.h [hunk 3]
```diff
@@ -971,10 +985,11 @@ class Object : public PointerValue {
...
 
-    virtual void privateFieldAdd(ExecutionState& state, AtomicString propertyName, const Value& value);
-    virtual void privateAccessorAdd(ExecutionState& state, AtomicString propertyName, FunctionObject* callback, bool isGetter, bool isSetter);
-    virtual Value privateFieldGet(ExecutionState& state, AtomicString propertyName);
-    virtual void privateFieldSet(ExecutionState& state, AtomicString propertyName, const Value& value);
+    void privateFieldAdd(ExecutionState& state, Object* contextObject, AtomicString propertyName, const Value& value);
+    void privateMethodAdd(ExecutionState& state, Object* contextObject, AtomicString propertyName, FunctionObject* fn);
+    void privateAccessorAdd(ExecutionState& state, Object* contextObject, AtomicString propertyName, FunctionObject* callback, bool isGetter, bool isSetter);
+    Value privateFieldGet(ExecutionState& state, Object* contextObject, AtomicString propertyName, bool shouldReferOuterClass = true);
+    void privateFieldSet(ExecutionState& state, Object* contextObject, AtomicString propertyName, const Value& value, bool shouldReferOuterClass = true);
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/runtime/Object.h:992]: Nice suggestion!

[the remark this answers is not in the document]
Comment (ksh8281) [reply src/runtime/Object.h]: I updated it

File: src/runtime/Object.h [hunk 4]
```diff
@@ -1172,6 +1187,14 @@ class Object : public PointerValue {
         return (ObjectRareData*)m_prototype;
     }
 
+    ObjectExtendedExtraData* extendedExtraData()
```
Comment (clover2123) [inline src/runtime/Object.h:1190]: It seems that this `extendedExtraData` function is defined but nowhere used.

File: src/codecache/CodeCacheReaderWriter.cpp [hunk 5]
```diff
@@ -800,6 +813,17 @@ InterpretedCodeBlock* CodeCacheReader::loadInterpretedCodeBlock(Context* context
             }
             codeBlock->rareData()->m_identifierInfoMap = varMap;
         }
+
+        if (size_t privateNameSize = m_buffer.get<size_t>()) {
+            AtomicStringTightVector* nameVector = new AtomicStringTightVector();
+            nameVector->resize(privateNameSize);
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/codecache/CodeCacheReaderWriter.cpp:819]: Nice suggestion!!

File: src/codecache/CodeCacheReaderWriter.cpp [hunk 6]
```diff
@@ -800,6 +813,17 @@ InterpretedCodeBlock* CodeCacheReader::loadInterpretedCodeBlock(Context* context
...
+            for (size_t i = 0; i < privateNameSize; i++) {
+                size_t stringIndex = m_buffer.get<size_t>();
+                auto name = m_stringTable->get(stringIndex);
+                nameVector->data()[i] = name;
```
[the remark this answers is not in the document]
Comment (ksh8281) [reply src/codecache/CodeCacheReaderWriter.cpp:823]: Nice suggestion!!

File: src/runtime/Object.h (hunk 4 above)
Comment (ksh8281) [reply src/runtime/Object.h:1190]: You are right!

**Comment by ksh8281** [comment]
I updated my pr

**Review by clover2123 — APPROVED** [review]
LGTM

**Review by bbrto21 — APPROVED** [review]

**Pull request merged**